File: common/src/main/java/com/ufsc/ine5418/protocol/Packet.java
package com.ufsc.ine5418.protocol;

import org.json.JSONObject;

import com.ufsc.ine5418.protocol.enums.HostType;
import com.ufsc.ine5418.protocol.enums.OperationType;
import com.ufsc.ine5418.protocol.enums.PayloadType;
import com.ufsc.ine5418.protocol.enums.Status;
import com.ufsc.ine5418.protocol.utils.FieldParser;
import com.ufsc.ine5418.utils.Logger;

public class Packet {

	private String host;
	private HostType hostType;
	private String token;
	private Status status;
	private OperationType operationType;
	private PayloadType payloadType;
	private JSONObject payload;

	public Packet(String host, HostType type, String token, Status status, OperationType operationType, PayloadType payloadType, JSONObject payload) {
		this.host = host;
		this.hostType = type;
		this.token = token;
		this.status = status;
		this.operationType = operationType;
		this.payloadType = payloadType;
		this.payload = payload;
	}

	public Packet(String jsonString) {
		JSONObject jsonPacket = new JSONObject(jsonString);

		try {
			this.host = jsonPacket.getString("host");
			this.hostType = HostType.valueOf(jsonPacket.getString("hostType"));
			this.token = FieldParser.nullableFieldToString(jsonPacket, "token");

			String status = FieldParser.nullableFieldToString(jsonPacket, "status");
			this.status = status != null ? Status.valueOf(status) : null;

			this.operationType = OperationType.valueOf(jsonPacket.getString("operationType"));
			this.payloadType = PayloadType.valueOf(jsonPacket.getString("payloadType"));
			this.payload = FieldParser.nullableFieldToJSONObject(jsonPacket, "payload");
		} catch (Exception e) {
			Logger.log(this.getClass().getSimpleName(), "Error parsing packet: " + e.getMessage());
		}
	}

	public String toString() {
		JSONObject jsonPacket = new JSONObject();

		jsonPacket.put("host", this.host);
		jsonPacket.put("hostType", this.hostType);
		jsonPacket.put("token", this.token != null ? this.token : JSONObject.NULL);
		jsonPacket.put("status", this.status != null ? this.status : JSONObject.NULL);
		jsonPacket.put("operationType", this.operationType);
		jsonPacket.put("payloadType", this.payloadType);
		jsonPacket.put("payload", this.payload != null ? this.payload : JSONObject.NULL);

		return jsonPacket.toString();
	}

}

File: client/source/packet.py
'''
Pacote.
'''

from json import dumps

from source.host_type import HostType
from source.status import Status
from source.operation_type import OperationType
from source.payload_type import PayloadType


class Packet():

    '''
    Pacote.
    '''

    _host: str
    _host_type: HostType
    _token: str | None
    _status: Status | None
    _operation_type: OperationType
    _payload_type: PayloadType
    _payload: dict | None

    def __init__(self,
                 host: str,
                 host_type: HostType,
                 operation_type: OperationType,
                 payload_type: PayloadType,
                 token: str | None = None,
                 status: Status | None = None,
                 payload: dict | None = None) -> None:
        self._host = host
        self._host_type = host_type
        self._token = token
        self._status = status
        self._operation_type = operation_type
        self._payload_type = payload_type
        self._payload = payload

    def __str__(self) -> str:
        packet = {
            'host': self._host,
            'hostType': self._host_type.value,
            'token': self._token,
            'status': self._status.value if self._status is not None else None,
            'operationType': self._operation_type.value,
            'payloadType': self._payload_type.value,
            'payload': self._payload
        }

        return dumps(packet)
